### PR TITLE
rng: implemented a random number generator

### PR DESCRIPTION
**Pull request by TheJJ**
It provides an easy and reliable way of reproducible random numbers,
along with platform-independent serialization.

Sanitized and updated version of #191. Thanks @schets !

### CODE REVIEW

File: cpp/rng/global_rng.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#define OPENAGE_RNG_GLOBAL_RNG_H_
+
+#include <cstdint>
+
+namespace openage {
+namespace rng {
+
+/**
+ * Returns the value of the global seed.
+ */
```
Comment (fr3akout) [inline cpp/rng/global_rng.h]: It would be nice to have single line comments if possible.
/** Returns ... */

**Comment by franciscod** [comment]
I've read "don't roll your own crypto" a lot of times. Isn't there any tested, used, "standard" rng for our use case?

anyway the name `murmurhash3` rocks

**Comment by TheJJ** [comment]
The advantage of this is that it's pretty easyly understandable, and supports very simple serialization. I think it's ready for merge, so :+1: etc.

**Comment by franciscod** [comment]
:+1:, the rng can be improved later if any crypto issues appear

**Comment by fr3akout** [comment]
I would like if somebody did a bit more research about the standard library of C++11, which already provides a pretty big set of random-utilities. If the standard library is sufficient for the use cases of openage, there is no need for this rng. Nevertheless this rng is not that critical, as it is not used for any crypto stuff at all.

**Comment by fr3akout** [comment]
I think this answers the serialization question: http://stackoverflow.com/questions/18361050/saving-random-number-generator-state-in-c11

File: cpp/rng/rng.h [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+#include <iostream>
+#include <string>
+
+namespace openage {
+namespace rng {
+
+/** \class rng
+ * This class holds an xorshift128+ rng, as well as
+ * utility generator functions
+ */
```
Comment (mic-e) [inline cpp/rng/rng.h]: why is this class name lower-case?

File: cpp/rng/rng.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		if (rnd.entropy() > 0){
+			rand1 = rnd() + (uint64_t(rnd()) << 32);
+		}
+	}
+	catch (...) {
+		throw util::Error(MSG(err) << "failed using std::random_device!");
+	}
+	return rand1;
+}
+
```
Comment (mic-e) [inline cpp/rng/rng.cpp]: How about seeding from `openage::timing::get_monotonic_time()`, `openage::timing::get_real_time()`, or, for added unpredictability, both?

File: cpp/rng/rng.cpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+ * murmurhash3 bit avalancher - output is uniformly distributed
+ * and changing one bit of input results in ~1/2 of output bits changed
+ * This function is a good way of mixing up an input,
+ * or mixing together two results with hash64(val1 ^ val2)
+ */
+static uint64_t hash64(uint64_t val) {
+	val ^= val >> 33;
+	val *= 0xff51afd7ed558ccd;
+	val ^= val >> 33;
+	val *= 0xc4ceb9fe1a85ec53;
```
Comment (mic-e) [inline cpp/rng/rng.cpp]: yay, magic numbers!
(maybe you could link the wikipedia article that explains the choice)

File: cpp/rng/rng.cpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+static uint64_t try_random_device() {
+	uint64_t rand1 = 0;
+	// std::random_device constructor may throw if it isn't available
+	try {
+		std::random_device rnd;
+		if (rnd.entropy() > 0){
+			rand1 = rnd() + (uint64_t(rnd()) << 32);
+		}
+	}
```
Comment (mic-e) [inline cpp/rng/rng.cpp]: I don't approve of your pokémon exception handling.

Please catch as `std::exception &e` and include `e.what()` in the MSG.

File: cpp/rng/rng_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+	int (*test_fnc)();
+	const char *name;
+};
+
+static void perform_tests(std::vector<data_test> tests) {
+	int ret;
+	for (data_test test : tests) {
+		if ((ret = test.test_fnc()) != -1) {
+			log::log(MSG(err) << "Testing " << test.name << " failed at stage " << ret);
+			throw "failed pairing heap tests";
```
Comment (mic-e) [inline cpp/rng/rng_tests.cpp]: sure, pairing heap tests...

File: cpp/rng/rng_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+	}
+	return 5;
+}
+
+static int rng_real_fill() {
+	constexpr size_t n = 1 << 7;
+	rng test{time_seed()};
+	rng test2{test};
+	double data[n];
+
```
Comment (mic-e) [inline cpp/rng/rng_tests.cpp]: I think `sizeof(data)` might be safer/have less redundancy.

**Comment by mic-e** [comment]
The code looks pretty nice and clean; in fact, I think it looks even cleaner due to the fact that the `rng` class has an all-lower-case name. Unfortunately, we agreed on upper-case letters in class names.

Since people often seem to associate RNGs with crypto, we should add a comment that states that this RNG was neither designed, nor tested, for use in crypto applications, and only morons would roll their own crypto.

File: cpp/rng/rng.cpp (hunk 3 above)
Comment (TheJJ) [inline cpp/rng/rng.cpp]: this is done in `time_seed`

File: cpp/rng/rng.cpp (hunk 5 above)
Comment (TheJJ) [inline cpp/rng/rng.cpp]: if you want your comments auto-outdate, you need to make it after the change you want.
The comment is associated with the 15 lines above it i think.

File: cpp/rng/rng.cpp (hunk 4 above)
Comment (mic-e) [inline cpp/rng/rng.cpp]: unfortunately the wikipedia article doesn't mention those magic numbers, it uses other ones...

**Pull request merged**